rust/src/quic/quic.rs

```diff
@@ -112,44 +122,111 @@ impl QuicState {
         return None;
     }
 
-    fn parse(&mut self, input: &[u8]) -> bool {
-        match QuicHeader::from_bytes(input, DEFAULT_DCID_LEN) {
-            Ok((rest, header)) => match QuicData::from_bytes(rest) {
-                Ok(data) => {
-                    // no tx for the short header (data) frames
-                    if header.ty != QuicType::Short {
-                        let mut sni : Option<Vec<u8>> = None;
-                        let mut ua : Option<Vec<u8>> = None;
-                        for frame in &data.frames {
-                            if let Frame::Stream(s) = frame {
-                                if let Some(tags) = &s.tags {
-                                    for (tag, value) in tags {
-                                        if tag == &StreamTag::Sni {
-                                            sni = Some(value.to_vec());
-                                        } else if tag == &StreamTag::Uaid {
-                                            ua = Some(value.to_vec());
-                                        }
-                                        if sni.is_some() && ua.is_some() {
-                                            break;
+    fn parse(&mut self, input: &[u8], to_server: bool) -> bool {
+        // so as to loop over multiple quic headers in one packet
+        let mut buf = input;
+        while buf.len() > 0 {
+            match QuicHeader::from_bytes(buf, DEFAULT_DCID_LEN) {
+                Ok((rest, header)) => {
+                    // unprotect
+                    if self.keys.is_none() && header.ty == QuicType::Initial {
+                        if let Some(version) = quic_rustls_version(u32::from(header.version)) {
+                            let keys = rustls::quic::Keys::initial(version, &header.dcid, false);
+                            self.keys = Some(keys)
+                        }
+                    }
+                    if header.length > rest.len() {
+                        //TODO event whenever an error condition is met
+                        return false;
+                    }
+                    let (next_buf, mut framebuf) = (&rest[header.length..], &rest[..header.length]);
+                    let mut rest2;
+                    if let Some(keys) = &self.keys {
```

Comment:
this should be broken out into a (set of) util func(s)

Reply:
Do you see the idiomatic rust way to do this ?

This being mainly 
- either use `input: &[u8]`
- or create a vector, and decode input into it, then use it as a slice

Reply:
I mean more the function from this point. It's very large

Reply:
I think I understand and agree, even before submitting the PR, but I am not sure I will be able to do it... because being in the same function helps with lifetime analysis (not sure about this naming)

Reply:
Managed something